pkg/chains/objects/objects.go

```diff
@@ -40,14 +43,158 @@ type Object interface {
 	runtime.Object
 }
 
-// Result is a generic key value store containing the results
+// GenericResult is a generic key value store containing the results
 // of Tekton operations. (eg. PipelineRun and TaskRun results)
-type Result struct {
+type GenericResult interface {
+	// GetName returns the name associated with the result.
+	GetName() string
+
+	// GetStringValue returns the string value of the result.
+	GetStringValue() string
+
+	// GetObjectValue returns the object value for the specified field.
+	GetObjectValue(field string) string
+
+	// ObjectValueIsNil checks if the object value is nil.
+	ObjectValueIsNil() bool
+}
+
+type GenericProvenance interface {
+	IsNil() bool
+	RefSourceIsNil() bool
+
+	GetRefSourceURI() string
+	GetRefSourceDigest() common.DigestSet
+	GetRefSourceEntrypoint() string
+
+	FeatureFlagsIsNil() bool
+	GetFeatureFlags() *config.FeatureFlags
+}
+
+// ProvenanceV1 is a struct implementing the GenericProvenance interface.
+type ProvenanceV1 struct {
+	Provenance *v1.Provenance
+}
+
+// RefSourceIsNil checks if the reference source is nil.
+func (p *ProvenanceV1) IsNil() bool {
+	return p.Provenance == nil
+}
+
+// RefSourceIsNil checks if the reference source is nil.
+func (p *ProvenanceV1) RefSourceIsNil() bool {
+	return p.Provenance.RefSource == nil
+}
+
+// GetRefSourceURI returns the URI of the reference source.
+func (p *ProvenanceV1) GetRefSourceURI() string {
+	return p.Provenance.RefSource.URI
+}
+
+// GetRefSourceDigest returns the digest set of the reference source.
+func (p *ProvenanceV1) GetRefSourceDigest() common.DigestSet {
+	return p.Provenance.RefSource.Digest
+}
+
+// GetRefSourceEntrypoint returns the entrypoint of the reference source.
+func (p *ProvenanceV1) GetRefSourceEntrypoint() string {
+	return p.Provenance.RefSource.EntryPoint
+}
+
+func (p *ProvenanceV1) FeatureFlagsIsNil() bool {
+	return p.Provenance.FeatureFlags == nil
+}
+
+func (p *ProvenanceV1) GetFeatureFlags() *config.FeatureFlags {
+	return p.Provenance.FeatureFlags
+}
+
+// ProvenanceV1Beta1 is a struct implementing the GenericProvenance interface.
+type ProvenanceV1Beta1 struct {
+	Provenance *v1beta1.Provenance
+}
+
+// RefSourceIsNil checks if the reference source is nil.
+func (p *ProvenanceV1Beta1) IsNil() bool {
+	return p.Provenance == nil
+}
+
+// RefSourceIsNil checks if the reference source is nil.
+func (p *ProvenanceV1Beta1) RefSourceIsNil() bool {
+	return p.Provenance.RefSource == nil
+}
+
+// GetRefSourceURI returns the URI of the reference source.
+func (p *ProvenanceV1Beta1) GetRefSourceURI() string {
+	return p.Provenance.RefSource.URI
+}
+
+// GetRefSourceDigest returns the digest set of the reference source.
+func (p *ProvenanceV1Beta1) GetRefSourceDigest() common.DigestSet {
+	return p.Provenance.RefSource.Digest
+}
+
+// GetRefSourceEntrypoint returns the entrypoint of the reference source.
+func (p *ProvenanceV1Beta1) GetRefSourceEntrypoint() string {
+	return p.Provenance.RefSource.EntryPoint
```

Comment:
v1beta1 can be [RefSource or ConfigSource](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#Provenance).

1. We should check that these are mutually exclusive
2. We should check both for provenance data.

Reply:
I am a bit confused by this.
Looking at https://github.com/tektoncd/pipeline/commit/ff4cf7ac7860655a86217d44d8dbc4131b76556d#diff-fa147d7e4cc13d0cdb1aabe655929d757d07c2a63eca242878376d85c48e0052, it doesn't seem that we've deprecated nor tomb-stoned the actual ConfigSource in v1beta1. But looks like [// Deprecated: Use RefSource instead ](https://github.com/tektoncd/pipeline/blob/c43d040d0954b152bd02878f7f55a43283e7441e/pkg/apis/pipeline/v1beta1/provenance.go#L23) indicates that we shall rely on RefSource as the source of truth?

Reply:
@wlynch I'm not sure I understand the suggestion here.  The v1beta1 Provenance object has ConfigSource and RefSource and ConfigSource was deprecated in v1 Provenance.  This code though doesn't change how ConfigSource is used IIUC, the RefSource vs ConfigSource is identical to how it works @ HEAD:

https://github.com/tektoncd/chains/blob/main/pkg/chains/objects/objects.go#L297-L302

https://github.com/tektoncd/chains/blob/main/pkg/chains/formats/slsa/attest/attest.go#L67

https://github.com/tektoncd/chains/blob/main/pkg/chains/formats/slsa/attest/attest.go#L111-L120

`attest.go` @ HEAD
```
func convertConfigSource(source *v1beta1.RefSource) slsa.ConfigSource {
	if source == nil {
		return slsa.ConfigSource{}
	}
	return slsa.ConfigSource{
		URI:        source.URI,
		Digest:     source.Digest,
		EntryPoint: source.EntryPoint,
	}
}
```

From the above snippets, currently tektoncd/chains @ HEAD only looks at RefSource information to populate slsa ConfigSource IIUC.  From what I understand from your comment this might be an issue but I think this is a separate bug and unrelated to this PR (this PR is keeping identical logic as to what is @ HEAD).  Happy to discuss more or understand your concern, IIUC the PR here keeps the old functionality